File: evaluate/covtype.py
import time
from river import forest, preprocessing, compose, metrics, drift, stream

DATA_PATH = '../data/covtype.data'
NUM_FEATURES = 54
SEED = 42
N_MODELS = 10
MAX_FEATURES = 5

def main():
    model = compose.Pipeline(
        preprocessing.StandardScaler(),
        forest.ARFClassifier(
            n_models=N_MODELS,
            max_features=MAX_FEATURES,
            seed=SEED,
            drift_detector=drift.binary.DDM(),
            warning_detector=drift.binary.DDM(drift_threshold=2.0, warning_threshold=1.0),
        )
    )

    metric = metrics.Accuracy()

    column_names = [f"feat_{i}" for i in range(NUM_FEATURES)] + ["label"]
    
    converters = {name: float for name in column_names[:-1]}
    converters["label"] = int

    start_time = time.time()
    
    count = 0

    for x, y in stream.iter_csv(
        DATA_PATH, 
        target="label", 
        fieldnames=column_names,  # <--- 这里改了
        converters=converters
    ):
        y = y - 1  
        y_pred = model.predict_one(x)
        metric.update(y, y_pred)
        model.learn_one(x, y)

    end_time = time.time()
    elapsed_ms = (end_time - start_time) * 1000

    print(f"Final Accuracy: {metric.get():.6f}")
    print(f"Time measured: {elapsed_ms:.0f} ms.")

if __name__ == "__main__":
    main()

File: evaluate/covtype.cpp
#include <cstdio>
#include <chrono>
#include <string>

#include "rivercpp/io/CSVReader.h"
#include "rivercpp/ARFClassifier.h"
#include "rivercpp/PipelineClassifier.h"
#include "rivercpp/StandardScaler.h"
#include "rivercpp/drift/DetectorConcept.h"
#include "rivercpp/drift/DDM.h"

constexpr int NUM_FEATURES = 54; 
constexpr int NUM_CLASSES = 7;

int main() {
    std::string data_path = "data/covtype.data";
    rivercpp::CSVReader reader(data_path, false);

    // DetectorFactory<ADWIN<5>, 0.01>, DetectorFactory<ADWIN<5>, 0.001>
    // DetectorFactory<DDM, 2.0>, DetectorFactory<DDM, 3.0>
    // DetectorFactory<HDDM_W, 0.005>, DetectorFactory<HDDM_W, 0.001>
    rivercpp::Classifier* model = new rivercpp::PipelineClassifier(
        new rivercpp::StandardScaler<NUM_FEATURES>(), 
        new rivercpp::ARFClassifier<NUM_FEATURES, NUM_CLASSES, 
            rivercpp::DetectorFactory<rivercpp::DDM, 2.0>, 
            rivercpp::DetectorFactory<rivercpp::DDM, 3.0> >(10, 5, 42) // (n_models, max_features, seed)
    );

    rivercpp::Accuracy<NUM_CLASSES> metric;

    auto begin = std::chrono::high_resolution_clock::now();

    while(reader.next()) {
        int pred = model->predict_one(reader.features);
        metric.update(reader.label-1, pred);
        model->learn_one(reader.features, reader.label-1);
    }

    auto end = std::chrono::high_resolution_clock::now();
    auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(end - begin);

    printf("Accuracy: %lf\n", metric.get());
    printf("Time measured: %lld ms.\n", static_cast<long long>(elapsed.count()));

    delete model;
    return 0;
}